types/src/multihash.rs

```diff
@@ -0,0 +1,214 @@
+use multihash_derive::{Hasher, MultihashDigest};
+use nmt_rs::simple_merkle::tree::MerkleHash;
+use nmt_rs::NamespaceMerkleHasher;
+
+use crate::byzantine::MULTIHASH_SHA256_NAMESPACE_FLAGGED_CODE;
+use crate::nmt::{
+    Namespace, NamespacedHashExt, NamespacedSha2Hasher, NAMESPACED_HASH_SIZE, NS_SIZE,
+};
+use crate::{Error, Result};
+
+pub type NamespacedHash = nmt_rs::NamespacedHash<NS_SIZE>;
+
+struct Sha256NamespaceFlaggedHasher {
+    hasher: NamespacedSha2Hasher,
+    hash: Option<NamespacedHash>,
+    hash_buf: [u8; NAMESPACED_HASH_SIZE],
+}
+
+impl Sha256NamespaceFlaggedHasher {
+    fn hash_data(&mut self, data: &[u8]) -> Result<()> {
+        let hash = if data.len() == NAMESPACED_HASH_SIZE * 2 {
+            let (left, right) = data.split_at(NAMESPACED_HASH_SIZE);
+            self.hasher
+                .hash_nodes(&left.try_into()?, &right.try_into()?)
+        } else {
+            self.hasher.hash_leaf(data)
+        };
+
+        self.hash = Some(hash);
+
+        Ok(())
+    }
+}
+
+impl Hasher for Sha256NamespaceFlaggedHasher {
+    fn update(&mut self, data: &[u8]) {
+        self.hash_data(data).expect("invalid data to hash");
+    }
+
+    fn finalize(&mut self) -> &[u8] {
+        let hash = if let Some(hash) = &self.hash {
+            hash
+        } else {
+            &NamespacedSha2Hasher::EMPTY_ROOT
+        };
+
+        self.hash_buf = hash.to_array();
+
+        &self.hash_buf
+    }
+
+    fn reset(&mut self) {
+        self.hash = None;
+    }
+}
+
+impl Default for Sha256NamespaceFlaggedHasher {
+    fn default() -> Self {
+        Self {
+            hasher: NamespacedSha2Hasher::with_ignore_max_ns(true),
+            hash: None,
+            hash_buf: [0u8; NAMESPACED_HASH_SIZE],
+        }
+    }
+}
+
+/// Outputs Multihashes, may panic when running `digest`. See `MultihashDigestExt` trait methods
+/// for safe alternative.
+#[derive(Clone, Copy, Debug, Eq, MultihashDigest, PartialEq)]
+#[mh(alloc_size = 128)]
+pub enum Code {
+    #[mh(code = MULTIHASH_SHA256_NAMESPACE_FLAGGED_CODE, hasher = Sha256NamespaceFlaggedHasher)]
+    Sha256NamespaceFlagged,
+}
+
+trait MultihashDigestExt {
+    fn digest_leaf(&self, ns: &Namespace, data: &[u8]) -> Result<Multihash>;
+    fn digest_inner(&self, left: &NamespacedHash, right: &NamespacedHash) -> Result<Multihash>;
+}
+
+impl MultihashDigestExt for Code {
+    fn digest_leaf(&self, ns: &Namespace, data: &[u8]) -> Result<Multihash> {
+        let namespaced_data = [ns.as_bytes(), data].concat();
+
+        Ok(self.digest(&namespaced_data))
+    }
+
+    fn digest_inner(&self, left: &NamespacedHash, right: &NamespacedHash) -> Result<Multihash> {
```

Comment:
can we do `digest_nodes`? I understand why digest_inner but it also can be understood as internal method that is used by others

Reply:
changed 👍